src/core/dep-graph.ts

```diff
@@ -175,6 +180,66 @@ class DepGraphImpl implements types.DepGraphInternal {
     };
   }
 
+  private nodeEquals(
+    graphA: types.DepGraphInternal,
+    nodeIdA: string,
+    graphB: types.DepGraphInternal,
+    nodeIdB: string,
+    compareRoot: boolean,
+    traversedNodes = new Set<string>(),
+  ): boolean {
+    // Skip root nodes comparision.
+    if (compareRoot || (nodeIdA !== graphA.rootNodeId && nodeIdB !== graphB.rootNodeId)) {
+      const pkgA = graphA.getNodePkg(nodeIdA);
+      const pkgB = graphB.getNodePkg(nodeIdB);
+
+      // Compare PkgInfo (name and version).
+      if (!_.isEqual(pkgA, pkgB)) {
+        return false;
+      }
+
+      const infoA = graphA.getNode(nodeIdA);
+      const infoB = graphB.getNode(nodeIdB);
+
+      // Compare NodeInfo (VersionProvenance and labels).
+      if (!_.isEqual(infoA, infoB)) {
+        return false;
+      }
+    }
+
+    let depsA = graphA.getNodeDepsNodeIds(nodeIdA);
+    let depsB = graphB.getNodeDepsNodeIds(nodeIdA);
+
+    // Number of dependencies should be the same.
+    if (depsA.length !== depsB.length) {
+      return false;
+    }
+
+    // Sort dependencies by name@version string.
+    const sortFn = (graph: types.DepGraphInternal) => (idA: string, idB: string) => {
+      const pkgA = graph.getNodePkg(idA);
+      const pkgB = graph.getNodePkg(idB);
+      return `${pkgA.name}@${pkgA.version}`.localeCompare(`${pkgB.name}@${pkgB.version}`);
+    };
+
+    depsA = depsA.sort(sortFn(graphA));
+    depsB = depsB.sort(sortFn(graphB));
+
+    // Compare Each dependency recursively.
+    for (let i = 0; i < depsA.length; i++) {
+      // Prevent cycles.
+      if (!traversedNodes.has(depsA[i])) {
```

Comment:
Maybe a small bug here? I think you should also make sure the `depsA[i]` == `depsB[i]` - what if A has a cycle, but B doesn't (or points in a different pkg)

Reply:
Good catch. I don't want to compare node ids. Will think about possible solution.